amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/Scheduler.java

```diff
@@ -1051,6 +1055,67 @@ private StreamIdentifier getStreamIdentifier(Optional<String> streamIdentifierSt
+    }
+
+    /**
+     * Create and return a copy of a {@link StreamConfig} object
+     * with {@link StreamIdentifier#streamArnOptional()} populated.
+     * Only to be used in multi-stream mode.
+     *
+     * @param streamConfig The {@link StreamConfig} object to return a copy of.
+     * @param kinesisRegion The {@link Region} the stream exists in, to be used for constructing the {@link Arn}.
+     * @return A copy of the {@link StreamConfig} with {@link StreamIdentifier#streamArnOptional()} populated.
+     */
+    private static StreamConfig withStreamArn(
+            @NonNull final StreamConfig streamConfig, @NonNull final Region kinesisRegion) {
+        Validate.isTrue(streamConfig.streamIdentifier().accountIdOptional().isPresent(),
+                "accountId should not be empty");
+        Validate.isTrue(streamConfig.streamIdentifier().streamCreationEpochOptional().isPresent(),
+                "streamCreationEpoch should not be empty");
+
+        log.info("Constructing stream ARN for {} using the Kinesis client's configured region - {}.",
+                streamConfig.streamIdentifier(), kinesisRegion);
+
+        final StreamIdentifier streamIdentifierWithArn = StreamIdentifier.multiStreamInstance(
+                constructStreamArn(
+                        kinesisRegion,
+                        streamConfig.streamIdentifier().accountIdOptional().get(),
+                        streamConfig.streamIdentifier().streamName()),
+                streamConfig.streamIdentifier().streamCreationEpochOptional().get());
+
+        return new StreamConfig(
+                streamIdentifierWithArn, streamConfig.initialPositionInStreamExtended(), streamConfig.consumerArn());
+    }
+
+    @RequiredArgsConstructor
+    private class StreamConfigMap extends ConcurrentHashMap<StreamIdentifier, StreamConfig> {
+        /**
+         * If {@link StreamIdentifier#streamArnOptional()} is present for the provided
+         * {@link StreamConfig#streamIdentifier()}, validates that the region in the stream ARN is consistent with the
+         * region that the Kinesis client ({@link RetrievalConfig#kinesisClient()}) is configured with.
+         * <p>
+         * In multi-stream mode, ensures stream ARN is always present by constructing it using the Kinesis client
+         * region when {@link StreamIdentifier#streamArnOptional()} is {@link Optional#empty()}.
+         * <p>
+         * {@inheritDoc}
+         */
+        @Override
+        public StreamConfig put(
+                @NonNull final StreamIdentifier streamIdentifier, @NonNull final StreamConfig streamConfig) {
+            final Region kinesisRegion = getKinesisRegion();
+
+            return super.put(streamIdentifier, streamConfig.streamIdentifier().streamArnOptional()
+                    .map(streamArn -> {
+                        Validate.isTrue(kinesisRegion.id().equals(streamArn.region().get()),
```

Comment:
Just realized that this is invoked in checkAndSyncStreamShardsAndLeases when new streams are discovered, when a shard sync task is generated and updates the currentStreamConfigMap. In that case if the validation fails, it can short-circuit the processing because the exception is only caught in runProcessLoop. This can prevent other streams from being processed. Ideally if there is an issue with a stream, its reasonable for Leader to log an error and skip the stream and maybe even emit a metric. But not let processing of other streams be affected. Can you fix it as a follow up ?